Monitoring_Tools_on_Graviton.md

```diff
@@ -0,0 +1,125 @@
+
+      10.961122204 seconds time elapsed
+
+#### Collect basic/ specific CPU hardware counters, for a specific command or system wide, for 10 seconds
+One can collect hardware events/ counters for an application, on a specific CPU, for a PID or system wide as follows:
+
+*$ perf stat -e cycles,instructions,cache-references,cache-misses,bus-cycles -a sleep 10*
+
+ Performance counter stats for 'system wide':
+
+         161469308      cycles                                                        (80.01%)
+         120685678      instructions              #    0.75  insn per cycle           (79.97%)
+          42132948      cache-references                                              (80.01%)
+           2001520      cache-misses              #    4.750 % of all cache refs      (80.02%)
+         160016796      bus-cycles                                                    (60.00%)
+
+      10.002896494 seconds time elapsed
+
+#### View the profile using perf report command
+*$ perf report*
+|Overhead  |   Command  |       Shared Object   |       Symbol|
+|---	|---	|---	|---    |
+|72.44%   |       functionA |       functionA       |       classA::functionA|
+|7.66%    |       functionB |       libB.so       |       classB::functionB|
+|...      |                 |                       |                 |
+|0.81%    |       functioA  |       libc-2.31.so    |       memcmp|
+
+More details on how to use Linux perf utility on AWS Graviton processors is available [here](https://github.com/aws/aws-graviton-getting-started/blob/main/optimizing.md#profiling-the-code).
+
+## Summary: Utilities on AWS Graviton vs. Intel x86 architectures
+|Processor	|x86	|Graviton2,3	|
+|---	|---	|---	|
+|CPU frequency listing	|*lscpu, /proc/cpuinfo, dmidecode*	|*dmidecode*	|
+|*turbostat* support	|Yes	|No	|
+|*hwloc* support	|Yes	|Yes	|
+|*lstopo* support	|Yes	|Yes	|
+|*i7z* Works	|Yes	|No	|
+|*lmbench*	|Yes	|Yes	|
+|Intel *MLC*  |Yes    |No     |
+|Performance monitoring tools	|_[VTune Profiler](https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html) and [PCM](https://github.com/opcm/pcm), [Linux perf](https://www.brendangregg.com/perf.html)_	|_[Linux perf](https://www.brendangregg.com/perf.html), [Arm Forge](https://developer.arm.com/Tools%20and%20Software/Arm%20Forge)_	|
+
+Utilities such as *lmbench* are available [here](http://lmbench.sourceforge.net/) and can be built for AWS Graviton processors to obtain latency and bandwidth stats.
+
+**Note:** On AWS Graviton 2/3 processors, Turbo isn’t supported. So, utilities such as ‘turbostat’ aren’t supported/ relevant for Arm architecture (and not on AWS Graviton processor either). Also, tools such as *[i7z](https://code.google.com/archive/p/i7z/)* for discovering CPU frequency, turbo, sockets and other information are only supported on Intel architecture/ processors. Intel *MLC* is a memory latency checker utility that is only supported on Intel processors.
```

Comment:
explictly call out the arm linux kernel community has decided not to put cpu frequency in /proc/cpuinfo which can be read by tools such as lscpu or directly?

Reply:
Ok, wasn't sure to put that in... I will add.